[docs] Improve the documentation covering react-router (#17343)
mui-org/material-ui: docs/src/pages/guides/composition/composition.md

{{"demo": "pages/guides/composition/Composition.js"}}

## Component prop

Material-UI allows you to change the root element that will be rendered via a prop called `component`.

### How does it work?

The component will render like this:

```js
return React.createElement(props.component, props)
```

For example, by default a `List` component will render a `<ul>` element.
This can be changed by passing a [React component](https://reactjs.org/docs/components-and-props.html#function-and-class-components) to the `component` prop.

> The following example will render the `List` component with a `<nav>` element as root element instead:

Reviewer: ```suggestion
The following example will render the `List` component with a `<nav>` element as the root element instead of `ul`:
```
?

Reviewer: Is this a good example? `<li>` aren't valid children of `<nav>` AFAIK

Reviewer 2: Right, adding `button` to make it a `nav > button`.

</List>
```

This pattern is very powerful and allows for great flexibility, as well as a way to interoperate with other libraries, such as your favorite routing or forms library.
But it also **comes with a small caveat!**

### Caveat with inlining

> Using an inline function as an argument for the `component` prop may result in **unexpected unmounting**, since you pass a new component to the `component` prop every time React renders.

Reviewer: ```suggestion
Using an inline function as an argument for the `component` prop may result in **unexpected unmounting**, since a new component is passed every time React renders.
```

`renderLink` will now always reference the same component.

> ### Caveat with props forwarding

Reviewer: ```suggestion
### Caveat with prop forwarding
```

> You can take advantage of the props forwarding to simplify the code.

Reviewer: ```suggestion
You can take advantage of the prop forwarding to simplify the code.
```

```

> ⚠️ However, this strategy suffers from a little limitation: properties collision.

Reviewer: ```suggestion
⚠️ However, this strategy suffers from a small limitation: property collisions.
```

> The component providing the `component` prop (e.g. ListItem) might not forward all its properties to the root element (e.g. dense).

Reviewer: ```suggestion
The component providing the `component` prop (e.g. ListItem) might not forward all the properties (for example dense) to the root element.
```
